File: apps/backend/app/auth/session_cache.py
import logging
from typing import Dict, Optional
from datetime import datetime
from app.models import AuthSession

logger = logging.getLogger("session-cache")

class AuthSessionCache:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(AuthSessionCache, cls).__new__(cls)
            cls._instance.cache: Dict[str, AuthSession] = {}
        return cls._instance

    def set(self, session: AuthSession) -> None:
        self.cache[session.project_id] = session
        logger.info(f"AuthSession cached for project {session.project_id}, expires at {session.expires_at}")

    def get(self, project_id: str) -> Optional[AuthSession]:
        session = self.cache.get(project_id)
        if not session:
            return None

        if datetime.utcnow() >= session.expires_at:
            logger.info(f"Cached AuthSession for project {project_id} expired. Evicting.")
            self.cache.pop(project_id, None)
            return None

        logger.info(f"Valid AuthSession retrieved from cache for project {project_id}")
        return session

    def invalidate(self, project_id: str) -> None:
        self.cache.pop(project_id, None)
        logger.info(f"AuthSession invalidated for project {project_id}")

auth_session_cache = AuthSessionCache()

File: packages/agents/src/auth/session-cache.ts
import type { AuthSession } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";

const logger = createLogger("session-cache");

/**
 * AuthSessionCache: In-memory session store with TTL expiry checks.
 * Prevents unnecessary re-authentication on consecutive test runs.
 */
export class AuthSessionCache {
  private static instance: AuthSessionCache;
  private cache: Map<string, AuthSession> = new Map();

  private constructor() {}

  public static getInstance(): AuthSessionCache {
    if (!AuthSessionCache.instance) {
      AuthSessionCache.instance = new AuthSessionCache();
    }
    return AuthSessionCache.instance;
  }

  /** Store an authenticated session in cache */
  public set(session: AuthSession): void {
    this.cache.set(session.projectId, session);
    logger.info({ projectId: session.projectId, expiresAt: session.expiresAt }, "AuthSession cached");
  }

  /** Retrieve valid, unexpired session for a project */
  public get(projectId: string): AuthSession | null {
    const session = this.cache.get(projectId);
    if (!session) return null;

    // Check if session has expired
    if (new Date() >= session.expiresAt) {
      logger.info({ projectId }, "Cached AuthSession expired. Evicting from cache.");
      this.cache.delete(projectId);
      return null;
    }

    logger.info({ projectId }, "Valid AuthSession retrieved from cache");
    return session;
  }

  /** Invalidate session for a project */
  public invalidate(projectId: string): void {
    this.cache.delete(projectId);
    logger.info({ projectId }, "AuthSession invalidated");
  }
}

export const authSessionCache = AuthSessionCache.getInstance();
